src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,191 @@
+	// Choose random height
+	unsigned long Index = static_cast<unsigned long>(m_Noise.IntNoise3DInt(a_AbsPos)) % m_TowerProperties.size();
+	do
+	{
+		Index = (Index + 1) % m_TowerProperties.size();
+		Properties = m_TowerProperties[Index];
+	} while (Properties.m_Height == 0);
+
+	// Make sure the tower does not cross chunk boarders
+	Pos.x = Clamp(Pos.x, Properties.m_Radius - 1, cChunkDef::Width - Properties.m_Radius);
+	Pos.z = Clamp(Pos.z, Properties.m_Radius - 1, cChunkDef::Width - Properties.m_Radius);
+
+	// Place obsidian pillar
+	for (int Y = 0; Y <= Properties.m_Height - 2; Y++)
+	{
+		for (int X = -Properties.m_Radius; X < Properties.m_Radius; X++)
+		{
+			for (int Z = -Properties.m_Radius; Z < Properties.m_Radius; Z++)
+			{
+				Vector3i NewPos = {Pos.x + X, Y, Pos.z + Z};
+				// The 3 was achieved by trial and error till the shape matched the notchian implementation
+				if ((NewPos - Vector3i(Pos.x, Y, Pos.z)).SqrLength() < Properties.m_Radius * Properties.m_Radius - 3)
+				{
+					a_ChunkDesc.SetBlockType(NewPos.x, NewPos.y, NewPos.z, E_BLOCK_OBSIDIAN);
+				}
+			}
+		}
+	}
+	// Spawn the bedrock
+	a_ChunkDesc.SetBlockType(Pos.x, Properties.m_Height - 1, Pos.z, E_BLOCK_BEDROCK);
+	// Spawn the Ender Crystal - the original passed position is not used because the tower might have been moved due to crossing chunk borders
+	const auto AbsPos = cChunkDef::RelativeToAbsolute(Pos, a_ChunkDesc.GetChunkCoords());
+	auto EnderCrystal = std::make_unique<cEnderCrystal>(AbsPos + Vector3d(0.5, Properties.m_Height, 0.5));
+	a_ChunkDesc.GetEntities().emplace_back(std::move(EnderCrystal));
+	// Spawn the fire
+	a_ChunkDesc.SetBlockType(Pos.x, Properties.m_Height, Pos.z, E_BLOCK_FIRE);
+	// Spawn the iron bars if there are some
+	if (Properties.m_HasCage)
+	{
+		// Place walls
+		for (int I = -2; I <= 2; I++)
+		{
+			a_ChunkDesc.SetBlockType(Pos.x - 2, Properties.m_Height - 1, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x - 2, Properties.m_Height, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x - 2, Properties.m_Height + 1, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + 2, Properties.m_Height - 1, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + 2, Properties.m_Height, Pos.z + I, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + 2, Properties.m_Height + 1, Pos.z + I, E_BLOCK_IRON_BARS);
+
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height - 1, Pos.z - 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height, Pos.z - 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height + 1, Pos.z - 2, E_BLOCK_IRON_BARS);
+
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height - 1, Pos.z + 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height, Pos.z + 2, E_BLOCK_IRON_BARS);
+			a_ChunkDesc.SetBlockType(Pos.x + I, Properties.m_Height + 1, Pos.z + 2, E_BLOCK_IRON_BARS);
+		}
```

Comment:
This is rather unintuitive code and also places multiple bars - it should be refactored.